### PR TITLE
Normalize test files in YAML

### PR DESCRIPTION
**Pull request by guillett**
* Amélioration technique.
* Zones impactées : `tests/**/*.{yaml, yml}`.
* Détails :
  - Normalisation des fichiers de tests en YAML
- - - -

L'objectif de cette PR est de faciliter la modification automatique/systématique des tests en YAML.
Pour :
- limiter les diffs lors du déplacement de la variable ASI de la famille à l'individu

Note :
Je sors du scope de cette PR la normalisation des fichiers YAML de paramètres.

Pour référence, un des objectifs était la préparation de ces fichiers à la manipulation(lecture, modification, suggestion d'intégration) par des systèmes informations (ie. legislation explorer contributif).

### CODE REVIEW

**Comment by guillett** [comment]
Je n'ai pas appliqué les modifications apportées par le script mais elles sont visibles sur le dernier commit de la branche [programmatic_tests_modif](https://github.com/openfisca/openfisca-france/tree/programmatic_tests_modif).

**Comment by guillett** [comment]
@openfisca/france-contributors que pensez-vous de cette PR ?

File: setup.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
 setup(
     name = 'OpenFisca-France',
-    version = '21.9.1',
+    version = '21.9.2   ',
```
Comment (benjello) [inline setup.py]: Espaces en trop.

**Comment by guillett** [comment]
Merci @fpagnoux. Il ne faut pas merger cette PR pour le moment. Je veux faire une passe studieuse pour éviter une modification automatique erronée.

**Comment by vinyll** [comment]
Shouldn't this been resolved and merged?

**Comment by guillett** [comment]
Merci @vinyll pour le ping. Je vais essayer de revenir sur cette PR aujourd'hui ou demain.

**Comment by guillett** [comment]
Je crois qu'on y est :)

Le mieux c'est de regarder https://github.com/openfisca/openfisca-france/pull/991/files?w=1 pour cacher les white-spaces-only changes.

**Comment by guillett** [comment]
En dépendance de https://github.com/etalab/biryani/pull/6

**Comment by guillett** [comment]
Avec [ce commit](https://github.com/openfisca/openfisca-france/pull/991/commits/a68cf8561f9da038e9dc3dfb7676b2df38d14a75) et surtout grâce à [une évolution demandée](https://bitbucket.org/ruamel/yaml/issues/158/incorrect-dumping-folded-scalars) sur ruamel.yaml il n'y a plus la dépendance à l'évolution sur biryani :-)

File: openfisca_france/scripts/yaml_round_trip.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+yaml.preserve_quotes = True
+yaml.width = 1000
+
+def roundTrip(path):
+    with io.open(path, 'r', encoding='utf8') as f:
+        try:
+            data = yaml.load(f)
+            with io.open(path, 'w', encoding='utf8') as f:
+                yaml.dump(data, f)
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (guillett) [reply openfisca_france/scripts/yaml_round_trip.py]: Thanks, I should not catch anything.

**Comment by guillett** [comment]
Many thanks @fpagnoux for your comments.

I have spent a lot of time on this PR and related ones (https://github.com/openfisca/openfisca-france/pull/1038, https://github.com/openfisca/openfisca-france/pull/1039, https://github.com/etalab/biryani/pull/6, https://bitbucket.org/ruamel/yaml/issues/216/round-trip-adds-trailing-spaces-in-file and https://bitbucket.org/ruamel/yaml/issues/158/incorrect-dumping-folded-scalars).

I am asking for more feedback and help from @openfisca/france-admin.

Personally, I am not in favor of using libraries in another language than Python (it is hard to maintain mes-aides-ui with NodeJs and Python).

## Performance-wise

Round trip check should be limited to files modified from either `origin/master` or the pull request base. Performance would than be proportional to the number of YAML file changed in the branch.

We can try to improve [ruamel.yaml](https://pypi.org/project/ruamel.yaml/).

## Workflow-wise

I agree it is not optimal and might be frustrating, linting/_round-triping_ should be more discoverable. However, the painful workflow you are describing is quite common for lint constraints.

On of the goals of that PR is to limit changes in YAML files to facilitate reviews and limit Git history (after manual or systematic editions).

We can try to make [ruamel.yaml](https://pypi.org/project/ruamel.yaml/) a Python equivalent of Prettier.

**Comment by fpagnoux** [comment]
>  the painful workflow you are describing is quite common for lint constraints

Not really: usually a linter runs locally in a couple of seconds. Having a linter who can only be run in CI and take 2 minutes is far from common.

But limiting the round-trip check to files that have changed relatively to `origin/master`, or to [the latest tagged parent commit](https://github.com/openfisca/openfisca-core/blob/master/.circleci/has-functional-changes.sh#L5) is probably the key to make the suggested implementation runable locally.

Also, as the linting is based on a round trip, I guess it should be easy to add a `make format-yaml` command that would automatically format the YAML, right?

**Comment by maukoquiroga** [comment]
Thanks for your efforts @guillett.

> I am asking for more feedback and help from @openfisca/france-admin.

> Personally, I am not in favor of using libraries in another language than Python (it is hard to maintain mes-aides-ui with NodeJs and Python).

Whatever the language of choice of the library or linter, if it aims for performance it'll invariably wrap a fast known library, like libyaml.

So it all depends on whether that fast library has a round trip encoder or not.

It is my knowledge that libyaml does not implement comment preservation at the moment, so ruaml.yaml implementation must be pure Python, which no surprise must be super slow.

I understand maintaining code on other languages than Python can add some overhead, but we're talking of different stacks: a linter is not a part of your application stack.

Also, if I understand correctly, if Prettier does the job you need, then the maintenance overhead is rather in the in-house than in the third-party-maintained code, Python or not. Whatever the language the linter is written with, there is not maintenance overhead if you can use it out-of-the-box, as it is not part of your application stack anyway.

> Not really: usually a linter runs locally in a couple of seconds. Having a linter who can only be run in CI and take 2 minutes is far from common.

👍 

> But limiting the round-trip check to files that have changed relatively to origin/master, or to the latest tagged parent commit is probably the key to make the suggested implementation runable locally.

For example `git merge-base HEAD master | xargs git diff --name-only HEAD | grep 'yaml\|yml'`.

**Comment by maukoquiroga** [comment]
Bonjour, un point d'étape sur cette issue :

> But limiting the round-trip check to files that have changed relatively to origin/master, or to the latest tagged parent commit is probably the key to make the suggested implementation runable locally.

J'ai inclus du linting limité au dernier `tag` ici #1135

> Also, as the linting is based on a round trip, I guess it should be easy to add a make format-yaml command that would automatically format the YAML, right?

Une fois fini #372, 

- [ ] Appliquer la même règle incrémental à ce linter
- [ ] Mettre à jour le `Makefile` pour faire pareil en local

Hors périmètre de cette PR mais à explorer :

- [ ] Voir comment un `format-yaml` pourrait prendre forme

---

Au-delà de ces éléments, je tiens à insister que si la même chose peut être accomplie avec [Prettier](https://prettier.io/), il faut qu'on trouve la solution la moins intensive en maintenance.

La réponse n'est pas univoque, c'est à décider ensemble :

- Solution node mais out-of-the-box maintenue par beaucoup de gens
- Solution Python mais house-made maintenue par votre humble serviteur

**Comment by guillett** [comment]
Bon, cette PR était une belle tentative, le coût de la normalisation et de la sauvegarde est trop élevée.

Cette PR n'était pas une finalité mais une étape intermédiaire du déplacement de l'ASI de la famille vers l'individu. Ce déplacement a pu être fait sans cette PR, cetter dernière peut être fermée.

**Pull request closed without merging**